x86: mmu: add ability for SoC to add MMU regions

dcpleung opened the pull request:
The SoCs usually have devices that are accessed through MMIO. This requires the corresponding regions to be marked readable and writable in the MMU or else accesses will result in page faults.

The UP Squared board is being used to make sure this works.

Fixes #25319

dcpleung committed:
x86: add ability for SoC to add MMU regions

The SoCs usually have devices that are accessed through MMIO.
This requires the corresponding regions to be marked readable
and writable in the MMU or else accesses will result in page
faults.

This adds a function which can be implemented in the SoC code to
specify those pages to be added to MMU.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: apollo_lake: add mmu regions for UARTs and I2Cs

UARTs and I2C controllers are accessed through MMIO and
these regions need to be added to MMU for proper access.

This also enable MMU for Apollo Lake by default since
serial console is now usable.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards: x86/up_squared: specify CONFIG_X86_MMU_PAGE_POOL_PAGES

Given that the UP Squared has relatively large memory, the default
number of pages allocated for page tables are not enough, and
resulting in asserting in the page table initialization code.
So change the number of pages to a large number to accomodate
various applications.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
serial: ns16550: return error when init fails

The init function returns successful even if the first
configuration function call fails. This may leave
a non-usable UART to be discoverable with
device_get_binding() which will definitely result
in lots of head scratching. So change the init function
to return properly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

jhedberg commented on boards/x86/up_squared/Kconfig.defconfig:
```diff
@@ -11,4 +11,7 @@ config BUILD_OUTPUT_STRIPPED
 	default y if !HPET_TIMER
 
+config X86_MMU_PAGE_POOL_PAGES
+	default 3072 if X86_MMU
```
When I boot my Up Squared I get: `Optimal CONFIG_X86_MMU_PAGE_POOL_PAGES 1033`

jhedberg replied:
With this PR, that is

nashif replied:
How much memory does you board have? I think this depends on the configuration, 2GB vs 4GB

dcpleung replied:
This number depends on how much memory is marked by the MMU code, which is dependent on the app size. A large number should be enough to cover most apps.

jhedberg commented:
I can confirm that this PR seems to work fine with the MMU support enabled, at least with all apps I tried so far. I can also confirm that user space works (tried with `samples/userspace/prod_consumer`).

jhedberg commented:
`samples/userspace/shared_mem` works as well

nashif commented:
> `samples/userspace/shared_mem` works as well

need to run the whole suite I guess

Commit:
tests/kernel/common: Make test_timeout_order 1cpu

This test works by starting a bunch of poll events, dropping the test
thread priority, calling k_poll(), and assuming that all the timeouts
that fired woke up high priority threads and thus ran before k_poll()
could return.  But that isn't true if you have another CPU that can
run the low priority thread while the last high priority thread
finishes up!

This just isn't SMP-correct.  Mark 1cpu.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

dcpleung committed:
boards: x86: make up_squared default to x86_64

This makes the up_squared board default to x86_64.
This also adds a new board, up_squared_32, for when 32-bit
is desired.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

Commit:
soc/x86/apollo_lake: IOAPIC ID's are not a SMP feature

This file consists only of an array of per-CPU IOAPIC ID's that
overrides the weak symbol defined by the architecture.

The IOAPIC IDs are only used when targetting a startup IPI for the
auxiliary right now, but the IDs are the IDs and represent hardware
truth.  They should be correct even if unused.

Using the wrong ones also breaks the tests/kernel/mp test, which calls
arch_start_cpu() when not in SMP mode as a deliberate unit test.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -817,6 +817,27 @@ static void add_mmu_region(struct x86_page_tables *ptables,
+		.flags = flags,
+	};
+
+	if (!user_table) {
```
are you sure this is right?
the code in `z_x86_paging_init()` calls add_mmu_region() on z_x86_user_ptables unconditionally for all MMU_BOOT_REGIONS


```
	Z_STRUCT_SECTION_FOREACH(mmu_region, rgn) {
		add_mmu_region(&z_x86_kernel_ptables, rgn, false);			add_mmu_region(&z_x86_kernel_ptables, rgn, false);
#ifdef CONFIG_X86_KPTI	#ifdef CONFIG_X86_KPTI
		add_mmu_region(&z_x86_user_ptables, rgn, true);			add_mmu_region(&z_x86_user_ptables, rgn, true);
#endif	#endif
	}		}
```

dcpleung replied:
Fixed this one.

dcpleung commented on drivers/serial/uart_ns16550_port_x.h:
```diff
@@ -13,7 +13,11 @@
+#ifndef UART_NS16550_PCIE_ENABLED
+const
+#endif
+struct uart_ns16550_device_config uart_ns16550_dev_cfg_@NUM@ = {
```
This is what I am planning to do, but it requires some thinking first. Hence the "workaround".

dcpleung replied:
Fixed.

dcpleung committed:
serial: ns16550: do not write to device cfg struct when PCIE=y

When PCIe is enabled for UART, the port address is probed during
initialization and is written back into the device config struct.
However, the device config struct is supposed to be const and
read only. This results in page faults when MMU is enabled as
the struct cannot be written into. So fix this by storing port
address in device data struct if a particular UART instance is
of PCIE.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented on boards/x86/up_squared/Kconfig.board:
```diff
@@ -2,5 +2,10 @@
+	bool "UP Squared (x86)"
 	depends on SOC_APOLLO_LAKE
+
+config BOARD_UP_SQUARED_64
```
I would make this board 64bit only, I do not think we should support 32bit by default.

andrewboie replied:
I think it would be good to eventually have two build targets, one 64bit and one for 32-bit HW testing since we have these on hand

andyross replied:
Split the difference and make the default "up_squared" board a 64 bit platform and have an "up_squared_32" for oddball uses?  I mostly agree with Anas that real applications on 64 bit CPUs are very rarely going to actually want to build a 32 bit target.

jhedberg replied:
I was about to suggest the same (i.e. up_squared_32)

andrewboie replied:
this sounds good to me

andyross reviewed (approved):
This all looks good to me.  Long term the policy of disallowing all MMIO memory by default is going to bite people doing board bringup again and again.  Maybe we need a layer like Linux has where drivers must all explicitly tell the kernel about their MMIO regions.  Then we could leave the page tables permissive (or leave paging disabled) during driver initialization and then abort (when CONFIG_ASSERT) when the final page tables don't match the registered regions.  The registration would then be a noop in production code.

dcpleung committed:
gpio: intel_apl: fix ISR function signature

Fix build error regarding the function signature of ISR.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: intel64: add missing _locore_end to linker script

_locore_end is missing from the linker script so add it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: acpi: make code 64-bit compatible

The integers used for pointer calculation were u32_t.
Change them to uintptr_t to be compatible with 64-bit.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
gpio: intel_apl: fix incorrect address access

The address used to access to REG_MISCCFG is not correct.
So fix it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andrewboie commented:
>  Maybe we need a layer like Linux has where drivers must all explicitly tell the kernel about their MMIO regions.

Yes, agreed, for VM the drivers will additionally need to know their mappings so I'm looking at driver interfaces which could accomplish this efficiently for both MMU and MPU use-cases.

andyross commented on drivers/serial/uart_ns16550.c:
```diff
@@ -733,10 +821,13 @@ static void uart_ns16550_isr(void *arg)
 {
 	struct device *dev = arg;
 	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
+	k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
```
Had some time to come back and look at this this morning.  Here is the deadlock: if the driver takes the lock before calling the registered handler, then the next uart action the handler does (for example: "try to read a byte") will deadlock, because the lock is already taken.

Just remove the locking from this spot, it's not needed.   Nothing is touching the hardware here, so there's nothing to protect.  Whatever the handler does already has its own locking.

jenmwms replied:
> yes, also my finding: [#23693 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/23693#discussion_r425886802)

This has been removed/resolved in the latest commit to #23693. Thanks for the feedback!

bwasim commented:
> Yes, agreed, for VM the drivers will additionally need to know their mappings so I'm looking at driver interfaces which could accomplish this efficiently for both MMU and MPU use-cases.

For targets that have an MMU, it should be the responsibility of the device driver to map / unmap the device memory space, something similar to `Mmap` function call in Linux (Ref: http://man7.org/linux/man-pages/man2/mmap.2.html).. 

However, the same cannot be easily done for MPU based targets as they have fixed number of regions in the hardware. For example, the Armv7-R MPU has 16 regions, and the Armv8-R has 24 regions in the hardware.

As the number of regions we want to map _may_ easily be more than regions supported in the hardware (for MPU), we'll have to support hot-swapping of the MPU regions for each implementation. Alternatively, for MPU, we can make the Mmap call a stub and make the BSP / SoC map all device memory in a single region at boot up (so no per-device mappings).. 

Though "hot swapping" of MPU regions is more flexible, it may be undesirable in some "real time" cases because the access to device that isn't mapped in the MPU will get a kernel exception and then the MPU in hardware will get re-programmed, which results in too much context switching.. For those cases, it probably makes more sense for MPU based targets to not support things like `Mmap` and just map all device memory in a single go.. This solution has its caveats also - Because all device memory is mapped by the kernel, dangling accesses to these memory won't be trapped / caught, leaving some bugs hidden.. But its simpler to implement and that's why I would prefer it..

andrewboie commented:
> However, the same cannot be easily done for MPU based targets as they have fixed number of regions in the hardware. For example, the Armv7-R MPU has 16 regions, and the Armv8-R has 24 regions in the hardware.

@bwasim We don't need to map MPU regions in the MPU, supervisor mode can access them already and the caching properties should already be set (PRIVDEFENA bit on ARM, etc).

We have supported MPUs for some time and have not had to set regions for them and don't plan to.

> device memory space, something similar to Mmap function call in Linux

It would be a lot closer to `ioremap()`.

andyross commented:
Am I right that the known uart failure was in tests/drivers/uart/uart_basic_api?  If so this bit fixes the problem for me when applied above this PR on up_squared (i.e. the new 64 bit MMU one, not the one in master):

```patch
diff --git a/drivers/serial/uart_ns16550.c b/drivers/serial/uart_ns16550.c
index c52643bb99..2098e7ee4b 100644
--- a/drivers/serial/uart_ns16550.c
+++ b/drivers/serial/uart_ns16550.c
@@ -841,13 +841,10 @@ static void uart_ns16550_isr(void *arg)
 {
        struct device *dev = arg;
        struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
-       k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
 
        if (dev_data->cb) {
                dev_data->cb(dev_data->cb_data);
        }
-
-       k_spin_unlock(&dev_data->lock, key);
 }
 
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
```

nashif commented:
> Am I right that the known uart failure was in tests/drivers/uart/uart_basic_api? If so this bit fixes the problem for me when applied above this PR on up_squared (i.e. the new 64 bit MMU one, not the one in master):

yes, also my finding: https://github.com/zephyrproject-rtos/zephyr/pull/23693#discussion_r425886802

dcpleung committed:
boards: x86/up_squared: remove SYS_CLOCK_HW_CYCLES_PER_SEC

This is defined in SoC and there is no need to override it
in the board config.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
timer: hpet: enable benchmarking during ISR

This adds the calls to read_timer_{start,end}_of_tick_handler()
to mark the start and end of ISR which will be used to display
the time spent in ISR with benchmarking tests.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented on boards/x86/up_squared/Kconfig.defconfig:
```diff
@@ -2,6 +2,26 @@
 if BOARD_UP_SQUARED
 
+config BOARD
+	default "up_squared_64"
```
this should be up_squared

nashif commented on boards/x86/up_squared/Kconfig.defconfig:
```diff
@@ -2,6 +2,26 @@
+if BOARD_UP_SQUARED_32
+
 config BOARD
 	default "up_squared"
```
and here it should be up_squared_32

jhedberg commented:
@dcpleung you'll need to rebase and fix the conflict since #25368 was merged
